File: solutions/0042-trapping-rain-water/trapping-rain-water.py
# Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.
#
#
# The above elevation map is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped. Thanks Marcos for contributing this image!
#
# Example:
#
#
# Input: [0,1,0,2,1,0,1,3,2,1,2,1]
# Output: 6
#


class Solution:
    def trap(self, height):
        '''
        solutin using two pointers
        '''
        if len(height) == 0: return 0
        l, max_l = 0, height[0]
        r, max_r = len(height) - 1, height[len(height) - 1]
        ans = 0
        while l < r:
            if max_l < max_r:
                ans += max_l - height[l]
                l += 1
                max_l = max(max_l, height[l])
            else:
                ans += max_r - height[r]
                r -= 1
                max_r = max(max_r, height[r])
        return ans
        

File: solutions/0042-trapping-rain-water/trapping-rain-water.cpp
// Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
//
//  
// Example 1:
//
//
// Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
// Output: 6
// Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.
//
//
// Example 2:
//
//
// Input: height = [4,2,0,3,2,5]
// Output: 9
//
//
//  
// Constraints:
//
//
// 	n == height.length
// 	0 <= n <= 3 * 104
// 	0 <= height[i] <= 105
//


class Solution {
public:
    int trap(vector<int>& height) {
        // height.push_back(INT_MAX);
        vector<int> stk;
        int ans = 0;
        for (int i = 0; i < height.size(); i++){
            int v = height[i];
            while (stk.size() && height[stk.back()] <= v){
                int tmp = stk.back();
                stk.pop_back();
                if (stk.size()){
                    ans += (min(v, height[stk.back()]) - height[tmp])*(i - stk.back() - 1);
                }
            }
            stk.push_back(i);
        }
        return ans;
    }
};
